security/rbac/engine/engine.go

```diff
@@ -0,0 +1,197 @@
+	return nil
+}
+
+// Decision is the enum type that represents different authorization
+// decisions a CEL engine can return.
+type Decision int32
+
+const (
+	// DecisionAllow indicates allowing the RPC to go through.
+	DecisionAllow Decision = iota
+	// DecisionDeny indicates denying the RPC from going through.
+	DecisionDeny
+	// DecisionUnknown indicates that there is insufficient information to
+	// determine whether or not an RPC call is authorized.
+	DecisionUnknown
+)
+
+// String returns the string representation of a Decision object.
+func (d Decision) String() string {
+	return [...]string{"DecisionAllow", "DecisionDeny", "DecisionUnknown"}[d]
+}
+
+// AuthorizationDecision is the output of CEL engine.
+// If decision is allow or deny, policyNames will either contain the names of
+// all the policies matched in the engine that permitted the action, or be
+// empty as the decision was made after all conditions evaluated to false.
+// If decision is unknown, policyNames will contain the list of policies that
+// evaluated to unknown.
+type AuthorizationDecision struct {
+	decision    Decision
+	policyNames []string
+}
+
+// Converts an expression to a parsed expression, with SourceInfo nil.
+func exprToParsedExpr(condition *expr.Expr) *expr.ParsedExpr {
+	return &expr.ParsedExpr{Expr: condition}
+}
+
+// Converts an expression to a CEL program.
+func exprToProgram(condition *expr.Expr) *cel.Program {
+	// TODO(@ezou): implement the conversion from expr to CEL program.
+	return nil
+}
+
+// Returns whether or not a policy is matched.
+func matchesCondition(condition *cel.Program, evalMap map[string]interface{}) (bool, error) {
+	// TODO(@ezou): implement the matching logic using CEL library.
+	return false, nil
+}
+
+// rbacEngine is the struct for an engine created from one RBAC proto.
+type rbacEngine struct {
+	action     pb.RBAC_Action
+	conditions map[string]*cel.Program
+}
+
+// Creates a new rbacEngine from an RBAC.
+func newRbacEngine(rbac pb.RBAC) rbacEngine {
+	action := rbac.Action
+	conditions := make(map[string]*cel.Program)
+	for policyName, policy := range rbac.Policies {
+		conditions[policyName] = exprToProgram(policy.Condition)
+	}
+	return rbacEngine{action, conditions}
+}
+
+// Returns whether a set of authorization arguments, as contained in evalMap, match an
+//  rbacEngine's conditions.
+// If any policy matches, the engine has been matched, and the list of matching policy
+//  names will be returned.
+// Else if any policy is missing attributes, the engine is not matched, and the list of
+//  policy names that can't be evaluated due to missing attributes will be returned.
+// Else, there is not a match.
+func (engine rbacEngine) matches(evalMap map[string]interface{}) (bool, []string) {
+	matchingPolicyNames := []string{}
+	unknownPolicyNames := []string{}
+	for policyName, condition := range engine.conditions {
+		match, err := matchesCondition(condition, evalMap)
+		if err != nil {
```

Comment:
I just checked the CEL code, and it turns out that the case for an unsuccessful evaluation is a bit different than we thought earlier: https://github.com/google/cel-go/blob/master/cel/program.go#L30-L43. So in general, CEL evaluation returned error doesn't mean "undecided". CEL evaluation returned error AND a nil val does.
I think that will make our `matchesCondition` return type a bit incompatible with the CEL. We might want to throw the error out in case of "Successful evaluation to an error result". WDYT?

Reply:
Yes, I agree. What exactly does "successful evaluation to an error result" mean though? Do you have any examples in mind?

Reply:
Also, this would likely be in a different part of the code, inside `matchesCondition` (which is currently implemented on the cel-integration branch but not in this PR). May I also assume that in the case of "successful evaluation to an error result", we don't do anything about the error?

Reply:
I am not entirely sure about the examples. It might be the case that the evaluations are able to be conducted, while there are some runtime errors, but yeah I couldn't come up with good examples for this.  @tristonianjones Could you please provide some inputs here?
My point here is that if the CEL evaluation engine could return some kind of errors other than attribute mismatching(for example, runtime error), we shouldn't hide these errors inside our engine code. That could make the testing and debugging harder.
@jiangtaoli2016 Can you please also take a look at this please?

Reply:
Just confirmed with Tristan and also verified by myself on my toy program.
 `Eval()` will output `ref.Val, *EvalDetails, error`.
- In the case of missing attributes, `ref.Val` will be an error string such as `no such attribute: id: 2, names: [source.address]`
- In the case of successful verification, `ref.Val` will be true or false.
- In the case of unsuccessful verification, `ref.Val` will be nil.

Reply:
Great!  Can I know what are the other two values(EvalDetails and error) in these three conditions?

Reply:
It sounds like in the case of missing attributes, error will be non-nil then? So the original logic of checking whether error is not nil seems to work?

Reply:
The `EvalDetails` is populated only if you want to trace expression evaluation. For your use case, I would expect `EvalDetails` to be ignored, and only `val` and `err` to be considered. If the `err` is populated, the evaluation failed either due to a precondition failure or a runtime issue in the expression. If the `val` is non-empty and `err` is empty this would point to a runtime expression evaluation issue, otherwise it's a precondition failure.

If you want to evaluate with partial state, the `types.Unknown` will come back as a `val`.
I'd recommend looking at the following for reference:
https://github.com/google/cel-go/blob/9b221121b9c8486869c73a8b52c9b32e28a7be44/cel/cel_test.go#L743

Reply:
if ref.Val == nil --> unsuccessful evaluation --> log error with `grpclog.Warning`
else if error != nil -> missing attributes
else -> successful evaluation

Reply:
SGTM.